### DISCUSSION TITLE
Berry tasmota module - no response from commands called from anonymous functions

### DISCUSSION THREAD

**Discussion by Grumium** [arendst/Tasmota#15304] [Q&A] [answered]
Hi all,
I experience a strange behavior when using Berry scripts together with the 'tasmota' module.

Please find my minimal working example below. I am expecting this to print "foo2" when POWER1 changes its state and I expect variable 'resp' to hold the response message. However 'resp' only returns 'nil'.
I suspect this has something to do with the anonymous function I define in the first rule. Is this behavior expected and how can I solve this?

```berry
class driver
    def init()
        tasmota.add_rule("POWER1#STATE", def (state) self.foo1(1,state) end)
        tasmota.add_rule("PWM", /->self.foo2())
    end
    def foo1(int,value)
        print("foo1", int, value)
        var resp = tasmota.cmd("PWM")
        print(resp)
    end
    def foo2()
        print("foo2")
    end
end
var d = driver()
tasmota.add_driver(d)
```
The console output was inconspicuous.

### Solutions

Based on answers by @sfromis this behavior can be avoided by using the `Event` command as shown in 
https://github.com/arendst/Tasmota/discussions/15304#discussioncomment-2515499

@s-hadinger suggested to defer the code by using `tasmota.set_timer(0, <func>)` which is what I went with.

**Comment by sfromis** [comment]
Read the Tasmota console at the same time to see what happens. You may be getting something like `{"Command":"Error"}`
That could happen if you did not configure for basic PWM control for the PWM command to apply. `SetOption15`. If you configure for basic PWM control, it will however no longer have a "Power" channel, if you want to have that, use `Dimmer` as a light control instead of `PWM`.

**Reply by Grumium** [reply]
I explicitly wrote "or any other command. PWM is just an example here.". SetOption15 is configured correctly and has nothing to do with the issue.

**Reply by sfromis** [reply]
Did you do as I suggested? Read the console. The PWM reference was just one possible cause, not meant as "here's *the* explanation". Hence "may ... something like".

**Reply by Grumium** [reply]
Yes, I did. I added the console output for completion above.

**Reply by sfromis** [reply]
When repeating your posted test shot using an `Event` command, I do see `foo2` being printed. 

BTW. Your edit to add details does not match your Berry code.

**Reply by sfromis** [reply]
Your added details still focuses **hard** on using PWM, not "any command". Your PWM rule **should not** match the JSON payload you get from the PWM command.

**Reply by Grumium** [reply]
> Your added details still focuses **hard** on using PWM, not "any command". Your PWM rule should not match the JSON payload you get from the PWM command.

I agree and I edited my inital post for an arbitrary command.

**Reply by sfromis** [reply]
You are still far from "any command". Your "TEST" command is improperly implemented, as of the docs, and your rule to match the outcome does still not match.

**Reply by Grumium** [reply]
I am obviously misunderstanding the working principle of events and triggers. I would be happy for any advice. The following code works. However, this is slightly different from the original problem, which did not use the "event" command.
```
class driver
    def init()
        tasmota.add_rule("POWER1#STATE", def (state) self.foo1(1,state) end)
        tasmota.add_rule("event#test", /s->self.foo2(1,s))
    end
    def foo1(int, state)
        tasmota.cmd("event test=" .. state)
    end
    def foo2(int,s)
        print("hello", int,s)
    end
end
var d = driver()
tasmota.add_driver(d)
```

**Reply by sfromis** [reply]
Ok, using `Event` is indeed different from wanting a `tasmota.add_rule` to hit based on the outcome of a command, as `Event` creates an explicit event to match the trigger. The suggestion from s-hadinger is a good one to avoid the race condition from calling further services from the handler for an add_rule. I'd recommend doing the set_timer to make it async.

Another technique also seems to work, doing the same as with "classic" Tasmota rules, where the outcome of commands issued by a rule in many cases is not available to further rule triggers, unless "wrapped" in a Backlog command.

**Comment by s-hadinger** [comment, marked as answer]
I think that running a command from within a rule creates some problems to get back the result. Actually running a command from within a command is not supported. The best way is to defer the code with `tasmota.set_timer(0, <func>)`

Like this (used `Status 4` as a trigger):
``` berry
class driver
    def init()
        tasmota.add_rule("StatusMEM", def (state) self.foo1(1,state) end)
        tasmota.add_rule("StatusMEM", /->self.foo2())
    end
    def foo1(int,value)
        tasmota.set_timer(0, def ()
            print("foo1", int, value)
            var resp = tasmota.cmd("Status 5")
            print("resp", resp)
            end
            )
    end
    def foo2()
        print("foo2")
    end
end
var d = driver()
tasmota.add_driver(d)
```

**Reply by Grumium** [reply]
I see! Deferring the code is an elegant solution. I will test this in a second.

**Reply by sfromis** [reply]
Well, the test code did work when I tested using an `Event` command, instead of the variations in edits done to demonstrate the issue without getting the basics about matching triggers right.

**Reply by Grumium** [reply]
Thanks to both of you for looking into this and clearing up my confusion. Both solutions work for me.
